### PR TITLE
stats, parallelism for event-schemas

### PR DESCRIPTION
**Pull request by bipin-rudder**
**Fixes** # (*issue*)

## Type of change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update
## Checklist:
- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added unit tests for the code
- [ ] I have made corresponding changes to the documentation

### CODE REVIEW

File: event-schema/event_schema.go [hunk 1]
```diff
@@ -168,7 +181,11 @@ func (manager *EventSchemaManagerT) updateEventModelCache(eventModel *EventModel
 	if !ok {
 		manager.eventModelMap[writeKey][eventType] = make(map[string]*EventModelT)
 	}
-	manager.eventModelMap[writeKey][eventType][eventIdentifier] = eventModel
+	if len(manager.eventModelMap[writeKey]) < eventModelLimit {
+		manager.eventModelMap[writeKey][eventType][eventIdentifier] = eventModel
+	} else {
+		stats.NewTaggedStat("eventSchemas.droppedEventModelsCount", stats.CountType, stats.Tags{}).Increment()
```
Comment (SumanthPuram) [inline event-schema/event_schema.go:187]: As per convention, we use snake case for stat names.

Comment (SumanthPuram) [reply event-schema/event_schema.go:187]: Let's return if we are not handling it?

File: event-schema/event_schema.go [hunk 2]
```diff
@@ -183,7 +200,12 @@ func (manager *EventSchemaManagerT) updateSchemaVersionCache(schemaVersion *Sche
 	if !ok {
 		manager.schemaVersionMap[eventModelID] = make(map[string]*SchemaVersionT)
 	}
-	manager.schemaVersionMap[eventModelID][schemaHash] = schemaVersion
+
+	if len(manager.schemaVersionMap[eventModelID]) < schemaVersionPerEventModelLimit {
+		manager.schemaVersionMap[eventModelID][schemaHash] = schemaVersion
+	} else {
+		stats.NewTaggedStat("eventSchemas.droppedSchemaVersionsCount", stats.CountType, stats.Tags{"eventModelID":eventModelID}).Increment()
```
Comment (SumanthPuram) [inline event-schema/event_schema.go:207]: Same as above.

File: event-schema/event_schema.go [hunk 3]
```diff
@@ -561,8 +583,10 @@ func getSchema(flattenedEvent map[string]interface{}) map[string]string {
 		if reflectType != nil {
 			schema[k] = reflectType.String()
 		} else {
-			schema[k] = "unknown"
-			pkgLogger.Errorf("[EventSchemas] Got invalid reflectType %+v", v)
+			if !(v == nil && !shouldCaptureNilAsUnknowns) {
+				schema[k] = "unknown"
+				pkgLogger.Errorf("[EventSchemas] Got invalid reflectType %+v", v)
```
Comment (SumanthPuram) [inline event-schema/event_schema.go:588]: We are handling the nil type here. So, should we use debug instead of error?

**Review by SumanthPuram — CHANGES_REQUESTED** [review]
Need to change stats naming convention.

File: event-schema/event_schema.go (hunk 3 above)
Comment (sivashanmukh) [reply event-schema/event_schema.go:588]: No. If its not nil we should ideally not get unknown. So if we got unknown, we should log it.

File: event-schema/event_schema.go (hunk 1 above)
Comment (sivashanmukh) [reply event-schema/event_schema.go:187]: Done

**Review by sivashanmukh — APPROVED** [review]
LGTM

**Review by gane5hvarma — APPROVED** [review]

**Review by SumanthPuram — APPROVED** [review]

**Pull request merged**